### DISCUSSION TITLE
Second level folding indicator showing wrong state after folding top level heading

### DISCUSSION THREAD

**Discussion by akurth** [minad/org-modern#195] [Q&A]
First of all thanks for this package; I’m a long time happy user.

Not sure if this is desired behaviour, but to me it looks like it shouldn’t be. Suppose the following org file structure:
```
* Heading 1
** Heading 2
- Item 1
- Item 2
```
Now unfold Heading 1
unfold Heading 2
fold Heading 1
re-unfold Heading 1

The items below Heading 2 are now folded (which is correct), but the folding indicator still shows "unfolded".

**Comment by minad** [comment]
Thanks for the report. I cannot reproduce the problem. Can you give me a more detailed recipe please (keypresses)? Which Emacs/Org version do you use?

**Reply by akurth** [reply]
* Org mode version 9.6.15 (release_9.6.15 @ /usr/local/share/emacs/30.0.50/lisp/org/)
* GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.41, cairo version 1.18.0) of 2024-04-28

I run emacs -Q and eval (load "~/.config/emacs/elpa/org-modern-20240428.1504/org-modern.el").

The recipe:
* Find file /tmp/test.org and enter the above content.
* M-x org-modern-mode
* Hit TAB twice on Heading 1 to fold and unfold
* Encounter the first problem: Heading 2 indicator shows unfolded symbol, but content is folded
* Hit TAB on Heading 2 to actually unfold content
* Hit TAB on Heading 2 again to fold
* Hit TAB on Heading 1 twice to fold and unfold
* Hit TAB on Heading 1 again
* Encounter the second problem: Heading 2 indicator shows folded symbol, but content is unfolded

**Reply by minad** [reply]
Okay thanks. I don't see any of these problems on stable Emacs 29.2. If you can debug this further, I would appreciate it.

**Reply by akurth** [reply]
Hm, I have the exact same behaviour with a freshly installed (snap): GNU Emacs 29.3 (build 2, x86_64-pc-linux-gnu, GTK+ Version 3.24.20, cairo version 1.16.0) of 2024-03-25.

**Reply by minad** [reply]
I have no idea about the problem. Maybe there is a difference in the Org versions. I use the Org nightly build, but you also observed this on Emacs master which should include a recent Org version. Hopefully someone else can reproduce this too. For now you can use `(setq org-modern-star 'replace)` until the feature becomes more stable.

**Reply by akurth** [reply]
I’m not very experienced in elisp, but maybe this could be a hint:
When debugging the STATE argument in `org-modern--cycle` it shows CHILDREN after hitting TAB on Heading 1. This seems to only update the current line, but not Heading 2.
If I change the function to unconditionally call `font-lock-flush`, it seems to work as expected.

**Comment by rnadler** [comment]
Not sure if it's related, but since the latest folding indicator change I'm seeing level 3 triangle icons displayed as "boxed" `23F5` and `23F7`.  It's only level 3, levels 1, 2, 4, and beyond display fine.

![image](https://github.com/minad/org-modern/assets/948978/76fc70d8-6842-4018-9cd3-f92698d59829)


org: 9.6.24
emacs: 29.3.50

**Reply by minad** [reply]
No. You can use another font, e.g., Iosevka or reconfigure org-modern-fold-stars.

**Comment by zhenhua-wang** [comment]
The built-in org in 29.3 does not work. Updating to the latest Org works for me...

**Reply by minad** [reply]
I see. So we should maybe only enable `fold` by default on newer Emacs or add a workaround. Do you have a suggestion for a possible workaround?

**Reply by zhenhua-wang** [reply]
I checked the org-mode versions. The built-in org in emacs 29.3 is "9.6.15" and the latest org is "9.7-pre". But I am not sure which exact org version works, and whether the newer emacs contains that... So I can't make any real suggestion.

**Reply by zhenhua-wang** [reply]
Looking at emacs master, it seems that the built-in for emacs 30 is still "9.6.15" for now. Checking `emacs-version` might not be a good approach. We probably can check `org-version` directly, but again I am not sure which org-version actually works... We probably should mark this folding-star as an experiment feature for now?

https://github.com/emacs-mirror/emacs/blob/master/lisp/org/org.el#L12

**Reply by ak-ionos** [reply]
I am using org 9.6.30, but the cycling behaviour has not been changed for a longer time anyway.

Maybe I wasn’t able to express the problem correctly. The org-cycle-hook – if called after a local state change (hitting TAB on a heading) – can pass one of the the states "folded", "children", or "subtree". In this case org-modern--cycle calls font-lock-flush for this heading line only. It seems obvious that the other (sub-heading) folding indicators stay unchanged, regardless of what content is actually visible.

**Reply by minad** [reply]
@ak-ionos @akurth But it seems there is no issue on the newest (development?) Org version. So something relevant must have changed and we can hopefully work around it here.

**Reply by akurth** [reply]
You are right, the current org-mode git version (9.7-pre) does work. Sorry for the noise.

**Reply by zhenhua-wang** [reply]
> I see. So we should maybe only enable `fold` by default on newer Emacs or add a workaround. Do you have a suggestion for a possible workaround?

Org 9.7 is out and it will be part of emacs 30. https://www.reddit.com/r/orgmode/comments/1d69t3a/org_mode_97_is_out/?utm_source=share&utm_medium=mweb3x&utm_name=mweb3xcss&utm_term=1&utm_content=share_button

We probably can check for version 30 now.
